File: Algorithms/Java/Sorting/QuickSort.java
//Alonso Ortiz - 25/Aug/2019

import java.io.*;
import java.math.*;
import java.text.*;
import java.util.*;
import java.util.regex.*;
import java.util.stream.*;

public class QuickSort {

    /*
     * 1. Implementing my version of the Quicksort Algorithm.
     */
    public static List<Integer> myQuicksort(List<Integer> ar) {
        if(ar.size() < 2) {
            return ar;
        } else {
            List<Integer> less = new ArrayList<Integer>();
            List<Integer> more = new ArrayList<Integer>();
            List<Integer> subList = new ArrayList<Integer>();
            Random ran = new Random();
            
            int indexPivot = ran.nextInt(ar.size());
            Integer pivot = ar.get(indexPivot);
            ar.remove(indexPivot);

            for(Integer element : ar) {
                if(element.compareTo(pivot) < 0) {
                    less.add(element);
                } else {
                    more.add(element);
                }
            }

            subList.addAll(myQuicksort(less));
            subList.add(pivot);
            subList.addAll(myQuicksort(more));
            return subList;
        }
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        List<Integer> unordered = new ArrayList<Integer>();


        for(int i = 0; i<=10; i++) {
            Random ran = new Random();
            int randomInt = ran.nextInt(20) + 1;
            unordered.add(randomInt);
        }

        System.out.println(unordered);

        List<Integer> ordered = myQuicksort(unordered);

        System.out.println(ordered);
        
    }
}

File: Algorithms/Python/Recursion/quicksort.py
#Alonso - 12/Oct/2018
                                                                                                                                   
import random

def main():
	print("My Quicksort - Sorted Array: ", myQuicksort([]))
	print("My Quicksort - Sorted Array: ", myQuicksort([3]))
	print("My Quicksort - Sorted Array: ", myQuicksort([1, 3]))
	print("My Quicksort - Sorted Array: ", myQuicksort([3, 1]))
	print("My Quicksort - Sorted Array: ", myQuicksort([3, 1, 4, 2]))
	print("My Quicksort - Sorted Array: ", myQuicksort([3, 1, 4, 3, 2]))
	print("My Quicksort - Sorted Array: ", myQuicksort([7, 2, 4, 8, 3, 5, 9, 1]))

def myQuicksort(arr):
	if len(arr) < 2:
		return arr
	else:
		#Changed the pivot from the first element to a random element of the array to make the algorithm more efficient. On average, it will take O(n log n) time.
		pivot = arr[random.randint(0, len(arr)-1)]
		less = []
		more = []

		arr.remove(pivot)

		for i in arr:
			if(i <= pivot):
				less.append(i)
			else:
				more.append(i)

		return myQuicksort(less) + [pivot] + myQuicksort(more)

main()      